File: jpp/lab2i3/java/GF.java
public class GF implements Field<GF>{
    private static int p = 1234577;
    public static GFOperations operations = new GFOperations();

    private int value;

    private int[] extendedEuclidean(int a, int b) {
        if (a == 0) {
            return new int[]{0, 1};
        } else {
            int[] result = extendedEuclidean(b % a, a);
            int x = result[1] - (b / a) * result[0];
            int y = result[0];
            return new int[]{x, y};
        }
    }

    public GF(int value) {
        this.value = value % p;
    }
    
    public GF inverse() {
        int[] result = extendedEuclidean(this.value, p);
        int x = result[0];
        return new GF((x + p) % p);
    }

    public void assign(GF other) {
        this.value = other.value;
    }

    //operacje arytmetyczne
    public void add(GF other) {
       this.value = (this.value + other.value) % p;
    }
    public void subtract(GF other) {
        this.value = (this.value - other.value + p) % p;
    }

    public void multiply(GF other) {
        this.value = (int)((long)this.value * (long)other.value % p) ;
    }

    public void divide(GF other) {
        if (other.value == 0) {
            throw new ArithmeticException("Division by zero");
        }
        this.multiply(other.inverse());
    }    

    //operatory porównania
    public boolean isValue(int value) {
        return this.value == GF.toField(value);
    }

    //konwersja int do GF
    public static int toField(int value) {
        while (value < 0) {
            value += p;
        }
        return value % p;
    }

    public int getValue() {
        return value;
    }

    public void setValue(int value) {
        this.value = value;
    }

    
    public static void setP(int newP) {
        p = newP;
    }

    public static int getP() {
        return p;
    }

    public String toString() {
        return "GF(" + Integer.toString(value)+ ")";
    }

}

File: jpp/lab2i3/python/Gf.py
class GF:
    p = 1234577

    @staticmethod
    def set_prime(prime):
        GF.p = prime

    @staticmethod
    def get_p():
        return GF.p

    def __init__(self, value):
        self.value = value % GF.p

    def __int__(self):
        return int(self.value)

    def __str__(self):
        return f"GF({self.value})"

    def __extended_euclid(self, a, b):
        if b == 0:
            return 1, 0
        else:
            x, y = self.__extended_euclid(b, a % b)
            return y, x - y * (a // b)

    def inverse(self):
        x, y = self.__extended_euclid(self.value, GF.p)
        return GF((x + self.p) % GF.p)

    def __add__(self, other):
        return GF((self.value + other.value) % GF.p)

    def __sub__(self, other):
        return GF((self.value - other.value) % GF.p)

    def __mul__(self, other):
        return GF((self.value * other.value) % GF.p)

    def __truediv__(self, other):
        if other.value == 0:
            raise ZeroDivisionError
        return self * other.inverse()

    # operatory porównania
    def __eq__(self, other):
        return self.value == other.value

    def __ne__(self, other):
        return self.value != other.value

    def __lt__(self, other):
        return self.value < other.value

    def __le__(self, other):
        return self.value <= other.value

    def __gt__(self, other):
        return self.value > other.value

    def __ge__(self, other):
        return self.value >= other.value
